### ISSUE TITLE
Primary Key in Output Mapping

### ISSUE THREAD

**Issue by ondrejhlavacek** [keboola/kbc-ui#681]
vycházím z https://github.com/keboola/kbc-ui/issues/674#issuecomment-251039691 a https://github.com/keboola/kbc-ui/issues/674#issuecomment-251042421

**Comment by ondrejhlavacek** [comment]
### Output Mapping Transformace 
1) Když tabulka neexistuje, musí uživatel zadat PK v output mappingu. 
2) Pokud tabulka již existuje, PK by nemělo jít změnit.

Co ale dělat v situaci, kdy PK je v transformaci nastavený, ale pak někdo přímo na tabulce změní PK? Co má ukazovat output mapping a jak se má chovat? Transformace začne ihned failovat.

### Output Mapping Komponent
Pokud komponenta zná svoje primární klíče, měla by je předat do UI (buď manifesty, nebo pojmenované mappingy). Opět může dojít k nekonzistenci a je potřeba to nějak v UI odkomunikovat a nechat uživatele opravit.

**Comment by ondrejhlavacek** [comment]
https://github.com/keboola/kbc-ui/issues/671

**Comment by Halama** [comment]
https://keboola.zendesk.com/agent/tickets/6071

Stejný problém u komponenty. V konfiguraci je nastvený PK ale tabulka ho z nějakého důvodu nastavený nemá. Docker-runner to ale zřejmě narozdíl od transfromací nekontroluje a neskončí?

**Comment by Halama** [comment]
https://github.com/keboola/docker-bundle/blob/63d656f0286f70aa4daaea382bfd38d0c0c63b04/Docker/StorageApi/Writer.php#L422

**Comment by Halama** [comment]
Už si vzpomínám že je to ve stavu kdy jenom posíláme warning ale kvůli kompatibilitě to nemůžeme zaříznout?

**Comment by Halama** [comment]
Jaký bude finální řešení, failne to když to nebude sedět nebo se to pokusí tabulku modifikovat?

**Comment by Halama** [comment]
U transformací už to failuje teď:
![image](https://cloud.githubusercontent.com/assets/903531/22463339/6d0f2154-e7b1-11e6-9e28-04d32741859d.png)

**Comment by ondrejhlavacek** [comment]
@Halama u transformací bylo třeba 60 výskytů nekozistence, tam jsem to ručně prošel u všech projektů a domluvil s nima vyřešení, a pak se nasadila striktní kontrola.

U docker runneru je ta situace o dost složitější. Každý den je tam cca 300+ nalezených výskytů nekonzistence, a to na dvou místech

a) v konfiguracích (output mapping vs reálnej stav)
b) v manifestech, co vygeneruje appka - uživatel změnil stav PK na cílový tabulce nebo appka (třeba i custom science) začala generovat jinej PK

To už není taková sranda na vyřešení. Nejde to detekovat v jeden okamžik z konfigurací, ale je potřeba sbírat data. 

Jestli tě napadá, jak z toho ven, tak beru jakýkoliv nápady - já to zatím uložil k ledu, protože to je zdánlivě úplná drobnost, ale nepodařilo se mi najít zatím žádný řešení, který by nebyl obrovskej aparát :-(

**Comment by Halama** [comment]
@ondrejhlavacek tu variantu ze by to tu tabulku modifikovalo podle nastaveni jsme asi zavrhli ze?

**Comment by ondrejhlavacek** [comment]
@Halama no, bál jsem se, že by to mohlo vyfailovat, když to najde duplicity, a tím by se zase mohly funkční orchestrace rozbít. Navíc bych očekával, že někdo bude mít 2 konfigurace, který zapisujou do stejný tabulky s jinejma PK (v transformacích jich bylo hafo).

Furt mi z toho vychází nejlíp validace projektu (na identifikaci postižených konfiguraci/tabulek) a pak nějakej termín, kdy zapnem striktní kontrolu a pak neustálá buzerace projektů, který to podle validace nefixnuli.

**Comment by ondrejhlavacek** [comment]
Ale jak nad tím teď přemýšlím, tak by se dala udělat neagresivní změna PK - když nad tím Storage vyfailuje, tak by docker runner jen hodil warning a pokračoval dál. Když by tohle nějakou chvíli běželo, vidělo by se, na kolika chybnejch konfiguracích by se to ustálilo...

**Comment by Halama** [comment]
to nezni spatne.

**Comment by ondrejhlavacek** [comment]
Jo, pustím se do toho ještě tenhle týden.
